crates/node-builder/src/exex.rs

```diff
@@ -0,0 +1,81 @@
+#![allow(dead_code)]
+
+use crate::{BuilderContext, FullNodeTypes};
+use futures::{future::BoxFuture, Future, FutureExt, Stream};
+use reth_primitives::BlockNumber;
+use std::pin::Pin;
+
+/// An ExEx (Execution Extension) that processes new blocks and emits events on a stream.
+pub trait ExEx: Stream<Item = ExExEvent> + Send + 'static {}
+
+/// Events emitted by an ExEx.
+#[derive(Debug)]
+pub enum ExExEvent {
+    /// Highest block processed by the ExEx.
+    ///
+    /// ExEx should guarantee that it will not require all earlier blocks in the future, meaning
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    FinishedHeight(BlockNumber),
+}
+
+/// Captures the context that an ExEx has access to.
+#[derive(Debug)]
+pub struct ExExContext<Node: FullNodeTypes> {
+    /// Different components of the node.
+    ///
+    /// Can be used to access the head block, the database, a stream of canonical state
+    /// notifications, etc.
+    pub builder: BuilderContext<Node>,
```

Comment:
if we're reusing this then this should be private and we add getter functions

but we're also deconstructing this type right away, so I think we should just flatten the fields here

https://github.com/paradigmxyz/reth/blob/c868c52063a7f507f2a455a752e159c9b5f08402/crates/node-builder/src/builder.rs#L511-L518

Reply:
yep, flattened